### ISSUE TITLE
Detect long reads in Nimble stack for BLE_GATT_ACCESS_OP_READ_CHR (IDFGH-10917)

### ISSUE THREAD

**Issue by western-hoolock** [espressif/esp-idf#12112]
### Is your feature request related to a problem?

I want to be able to distinguish between a normal read and a long read in the GATT server callback for the ```BLE_GATT_ACCESS_OP_READ_CHR``` operation. Nimble takes care of splitting up characteristics that a are larger than the MTU in multiple packets but it does so by calling the callback with ```BLE_GATT_ACCESS_OP_READ_CHR```multiple times without any flag to check whether the current read operation is a partial one. This causes a problem in my case because the characteristic gets updated in the background and I need to know if a long read operation is still in progress or else the characteristic will contain a mix of old and new data.

### Describe the solution you'd like.

```
int gatt_service_callback(uint16_t conn_handle, uint16_t attr_handle, struct ble_gatt_access_ctxt *ctxt, void *arg)
{
    switch (ctxt->op) {
    case BLE_GATT_ACCESS_OP_READ_CHR:
        // I want to be able to check a flag here to determine if this is a single read or part of a long read.
      
...
```

### Describe alternatives you've considered.

I've seen some code in a BLE library for Arduino that does this check:
```
if (ctxt->om->om_pkthdr_len > 8) {
    // This is a single read or start of a long read, make local copy.
    ...
}
```

If om_pkthdr_len > 8 then it's a single read/first read of a long read, which means I should create a local copy and use that for any reads that do not satisfy that check. This seems to work but I'm not even sure what exactly it's checking or if it's guaranteed to work between Nimble versions. Feels hacky.

**Comment by ESPAbhinav** [comment]
Hi @western-hoolock,
-  We have checked your issue and it is observed that there is no direct flag to determine if the current read is a single read or part of a long read in the server application callback however same can be checked on the client side.
-  As per the upstream Nimble implementation, the stack sends the read response itself to the client and hence there is no response by the server application. I have suggested the following way which you can apply to the server application and determine whether the current read is a single read or part of a long read.
The following are suggested changes.
```
static int cnt = 0;

static int gatt_svc_access(uint16_t conn_handle, uint16_t attr_handle,
                struct ble_gatt_access_ctxt *ctxt, void *arg)
{
    const ble_uuid_t *uuid;
    int rc;
    int mtu;

    switch (ctxt->op)
    {
    case BLE_GATT_ACCESS_OP_READ_CHR:
        
        if (attr_handle == gatt_svr_chr_val_handle)
        {
            rc = os_mbuf_append(ctxt->om,
                                gatt_svr_chr_val,
                                sizeof(gatt_svr_chr_val));

            mtu = ble_att_mtu(conn_handle);
            
            // ---------------------Suggested changes -------------------------------//
            
            cnt++;
            if (cnt == (((sizeof(gatt_svr_chr_val) / (mtu - 1)) + 1)))
            {
                printf("Current read is the last read of long read or single read\n");
                cnt = 0;
            }
            else
            {
                printf("Current read is part of long read\n");
            }
            return rc == 0 ? 0 : BLE_ATT_ERR_INSUFFICIENT_RES;
        }
```
In the above code, we have added one counter which will increase for every callback with `BLE_GATT_ACCESS_OP_READ_CHR`. Total number of times callback will hit can be calculated as               
`size of (char_val)/(mtu-1)` as offset is increasing by mtu-1.

- Nimble stack takes care of splitting up characteristics that are larger than the MTU in multiple packets which invokes the same callback ( for both normal read and long read) with BLE_GATT_ACCESS_OP_READ_CHR where only the characteristic value is appended. 

Below is the patch file attached for the suggested changes.

[0001-Server-application-changes-for-gattt_svc_acsess-call.zip](https://github.com/espressif/esp-idf/files/12496576/0001-Server-application-changes-for-gattt_svc_acsess-call.zip)

**Comment by western-hoolock** [comment]
Thanks, I wasn't aware of the ```ble_att_mtu``` function and the fact that ```ble_att_mtu - 1``` is the available space for the characteristic.
While the suggested changes work I'm still curious if:
```if (ctxt->om->om_pkthdr_len > 8)``` is a reliable check to know if this is the start of a read operation? Since long reads always seem to have ```om_pkthdr_len set to <= 8``` that check can potentially be used to know when to "latch" my characteristic's data.

**Comment by ESPAbhinav** [comment]
Thanks for confirming that the suggested changes work. 
- In a nimble stack, the packet header length is calculated as 
` pkthdr_len = user_pkthdr_len + sizeof(struct os_mbuf_pkthdr) `  which turns out to be 8 bytes. So for both the operation  (read and long read )  `ctxt->om->om_pkthdr_len` will give the value of 8. I have verified it by the runtime console log. 
- So relying on packet header length may not be helpful in our case. Below is the definition of a struct `os_mbuf_pkthdr`
```c
 struct os_mbuf_pkthdr {                   
     /**
      * Overall length of the packet.
      */
     uint16_t omp_len;
     /**
      * Flags
      */                
     uint16_t omp_flags;
     STAILQ_ENTRY(os_mbuf_pkthdr) omp_next;
 };
```
We can see that the size of the above definition is 8 bytes. Also if we want to check the first read of the operation i am suggesting adding one extra condition in previously suggested changes 
```c
// ---------------------Suggested changes -------------------------------//
            
            cnt++;
            if (cnt == (((sizeof(gatt_svr_chr_val) / (mtu - 1)) + 1)))
            {
                printf("Current read is the last read of long read or single read\n");
                cnt = 0;
            }
            else if (cnt == 1)
            {   // Extra Condition 
               printf("This is start of read operation for both read and long read \n");
            }
            else
            {
                printf("Current read is part of long read\n");
            }
            return rc == 0 ? 0 : BLE_ATT_ERR_INSUFFICIENT_RES;

**Comment by rahult-github** [comment]
Closing issue. Feel free to reopen in case needed.

**Comment by vandy** [comment]
@western-hoolock, I'm not sure that it's the nimble stack which splits an attribute value reading to several callbacks. As I see it, when a client sends `ATT_READ_REQ` it gets response from the server which contains the attribute data. If the data length is the maximum of a packet (MTU, actually MTU - 1) than the client could decide to send `ATT_READ_BLOB_REQ` to try to read the remained data until it gets no data. This is a long attribute read.

I think `ctxt->om->om_pkthdr_len > 8` won't work in most cases, because of https://github.com/apache/mynewt-nimble/blob/master/nimble/host/src/ble_gatts.c#L368. When a client does blob read (sending offset value larger than 0), the nimble stack creates packet header mbuf (which contains packet headers). So `om->om_pkthdr_len` should be almost always greater than 8. The application callback will interact with this mbuf.

One of the approaches to indicate not the first read is passing an offset to the application callback. I've made a [PR](https://github.com/apache/mynewt-nimble/pull/1846) to the original repository.

**Comment by AxelLin** [comment]
> Hi [@western-hoolock](https://github.com/western-hoolock),
> 
> * We have checked your issue and it is observed that there is no direct flag to determine if the current read is a single read or part of a long read in the server application callback however same can be checked on the client side.

Hi @rahult-github @ESPAbhinav 

Can you apply this:
https://github.com/apache/mynewt-nimble/pull/1846/commits/019c6e2eb332098f6846af17785596a02491cb7b
You can find discussions here: https://github.com/apache/mynewt-nimble/issues/1090

**Comment by ESPAbhinav** [comment]
Hi @AxelLin , Thanks for suggesting changes. We will go through the discussions and suggested changes.

**Comment by AxelLin** [comment]
> Hi [@AxelLin](https://github.com/AxelLin) , Thanks for suggesting changes. We will go through the discussions and suggested changes.

Hi @ESPAbhinav 

I still cannot find this fix (apache/mynewt-nimble@019c6e2)  in esp-nimble.
Also please consider this as bug fix for esp-idf release branches.
(The fix is so trivial, without the fix it needs much more complex logic in application to detect long reads.)
